pkg/store/proxy_heap.go

```diff
@@ -0,0 +1,743 @@
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+	"github.com/thanos-io/thanos/pkg/tracing"
+)
+
+// dedupResponseHeap is a wrapper around ProxyResponseHeap
+// that removes duplicated identical chunks identified by the same labelset and checksum.
+// It uses a hashing function to do that.
+type dedupResponseHeap struct {
+	h *ProxyResponseHeap
+
+	responses []*storepb.SeriesResponse
+
+	previousResponse *storepb.SeriesResponse
+	previousNext     bool
+}
+
+func NewDedupResponseHeap(h *ProxyResponseHeap) *dedupResponseHeap {
+	return &dedupResponseHeap{
+		h:            h,
+		previousNext: h.Next(),
+	}
+}
+
+func (d *dedupResponseHeap) Next() bool {
+	d.responses = d.responses[:0]
+
+	// If there is something buffered that is *not* a series.
+	if d.previousResponse != nil && d.previousResponse.GetSeries() == nil {
+		d.responses = append(d.responses, d.previousResponse)
+		d.previousResponse = nil
+		d.previousNext = d.h.Next()
+		return len(d.responses) > 0 || d.previousNext
+	}
+
+	var resp *storepb.SeriesResponse
+	var nextHeap bool
+
+	// If buffered then use it.
+	if d.previousResponse != nil {
+		resp = d.previousResponse
+		d.previousResponse = nil
+	} else {
+		// If not buffered then check whether there is anything.
+		nextHeap = d.h.Next()
+		if !nextHeap {
+			return false
+		}
+		resp = d.h.At()
+	}
+
+	// Append buffered or retrieved response.
+	d.responses = append(d.responses, resp)
+
+	// Update previousNext.
+	defer func(next *bool) {
+		d.previousNext = *next
+	}(&nextHeap)
+
+	if resp.GetSeries() == nil {
+		return len(d.responses) > 0 || d.previousNext
+	}
+
+	for {
+		nextHeap = d.h.Next()
+		if !nextHeap {
+			break
+		}
+		resp = d.h.At()
+		if resp.GetSeries() == nil {
+			d.previousResponse = resp
+			break
+		}
+
+		lbls := resp.GetSeries().Labels
+		lastLbls := d.responses[len(d.responses)-1].GetSeries().Labels
+
+		if labels.Compare(labelpb.ZLabelsToPromLabels(lbls), labelpb.ZLabelsToPromLabels(lastLbls)) == 0 {
+			d.responses = append(d.responses, resp)
+		} else {
+			// This one is different. It will be taken care of via the next Next() call.
+			d.previousResponse = resp
+			break
+		}
+	}
+
+	return len(d.responses) > 0 || d.previousNext
+}
+
+func (d *dedupResponseHeap) At() *storepb.SeriesResponse {
+	if len(d.responses) == 0 {
+		return nil
+	} else if len(d.responses) == 1 {
+		return d.responses[0]
+	}
+
+	chunkDedupMap := map[uint64]*storepb.AggrChunk{}
+
+	for _, resp := range d.responses {
+		if resp.GetSeries() == nil {
+			continue
+		}
+		for _, chk := range resp.GetSeries().Chunks {
+			for _, field := range []*storepb.Chunk{
+				chk.Raw, chk.Count, chk.Max, chk.Min, chk.Sum, chk.Counter,
+			} {
+				if field == nil {
+					continue
+				}
+				h := xxhash.Sum64(field.Data)
```

Comment:
As a follow up improvement, would it make sense to do this hashing in the stores so that we offload some work from the querier proxy? We might also think about other operations that can be pre-computed at the store level.